File: 002AddTwoNumbers.cpp
/*
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode *res, *q, *p;
        if(l1 == NULL)
        	return l2;
        else if (l2 == NULL)
        	return l1;
        else{
        	carry += l1->val + l2->val;	 	//first node
        	res = new ListNode(carry%10);	//assign space and calculate the value of first node 
        	carry /= 10;
        	l1 = li->next;
        	l2 = l2->next;
        	q = res;			//q always points to the last node in the result list
        }
        while(l1 != NULL && l2 != NULL){
        	carry += l1->val + l2->val;
        	p = new ListNode(carry%10);
        	carry/=10;
        	q->next = p;
        	q = p;
        	l1 = l1->next;
        	l2 = l2->next;
        }
        // if l1 has remaing nodes, direct handling
        while(l1 != NULL){
        	carry += l1->val;
        	p = new ListNode(carry%10);
        	carry /= 10;
        	q -> next = p;
        	q = p;
        	l1 = l1->next;
        }
        // if l2 has re,aomgmpdes, direct handling
        while(l2 != NULL){
        	carry += l2->val;
        	p = new ListNode(carry%10);
        	carry /= 10;
        	q->next = p;
        	q = p;
        	l2 = l2-> next;
        }
        // the final carry value maybe greater than 10, special handing
        while(carry!= 0){	
        	p = new ListNode(carry%10);
        	carry/= 10;
        	q->next = p;
        	q = p;
        }
        return ans;
    }

};

File: 002AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        #add a flag, added each position, if greater than 9, flag is 1,next position will add flag
        #when the last position flag is 1, it needs add 1 into the tail
        if l1 is None:
        	return l2
        if l2 is None:
        	return l1
        
        temp = ListNode(0)
        res = temp
        flag = 0
        while l1 or l2:
        	tmpsum = 0
        	if l1:
        		tmpsum = l1.val
        		l1 = li.next
        	if l2: 
        		tmpsum += l2.val
        		l2 = l2.next
        	tmpres = ((tmpsum + flag) %10)
        	flag = ((tmpsum + flag) // 10)
        	res.next = ListNode(tmpres)
        	res = res.next
        if flag:
        	res.next = ListNode(1)
        res = temp.next
        del temp
        return res


        '''
        if not l1 or not l2:
        	return l1 or l2
        root, add = l1, 0
        while l1 and l2:
        	cur = l1
        	add, cur.val = divmod(l1.val + l2.val + add, 10)
        	l1,l2 = l1.next, l2.next
        cur.next = l1 or l2
        while add:
        	if cur.next:
        		cur = cur.next
        		add, cur.val = divmod(cur.val + add, 10)
        	else:
        		cur.next = ListNode(1)
        		break
        return root
        '''



        